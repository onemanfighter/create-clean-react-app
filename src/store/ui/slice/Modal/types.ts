export enum ModalID {
  NONE = 'none',
  SEARCH = 'search',
}

export type ModalState = {
  modalID: ModalID;
  modalOpenState: ModalOpenState;
  modalData?: ModalData;
};

export type ModalData = {
  onModalClose: () => void;
};

export enum ModalOpenState {
  OPEN = 'OPEN',
  CLOSE = 'CLOSE',
}

export interface ModalStateAction {
  openModal: (id: ModalID, onModalClose: () => void) => void;
  resetModalState: () => void;
}

export type ModalStateSlice = ModalState & ModalStateAction;
