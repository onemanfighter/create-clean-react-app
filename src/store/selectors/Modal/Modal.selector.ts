import { AppStoreState } from '@store';

const modalSelector = (state: AppStoreState) => ({
  openState: state.Modal.modalOpenState,
  modalID: state.Modal.modalID,
  onModalClose: state.Modal.onModalClose,
  openModal: state.Modal.openModal,
  closeModal: state.Modal.closeModal,
});

export { modalSelector };
