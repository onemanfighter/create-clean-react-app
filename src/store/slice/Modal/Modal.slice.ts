import { AppStoreSlice } from '../../appStore/types';
import { ModalState, ModalStateSlice, ModalOpenState, ModalIDs } from './types';

const initialAlertState: ModalState = {
  modalID: undefined,
  onModalClose: undefined,
  modalOpenState: ModalOpenState.CLOSE,
};

const createModalSlice: AppStoreSlice<ModalStateSlice> = (set) => ({
  ...initialAlertState,
  openModal: (modalData) => {
    set((state) => {
      state.Modal.modalID = modalData.modalID;
      state.Modal.onModalClose = modalData.onModalClose;
      state.Modal.modalOpenState = ModalOpenState.OPEN;
    });
  },
  closeModal: () => {
    set((state) => {
      state.Modal.modalOpenState = ModalOpenState.CLOSE;
    });
    setTimeout(() => {
      set((state) => {
        state.Modal.modalOpenState = initialAlertState.modalOpenState;
      });
    }, 500 /* 0.5s */);
  },
});

export default createModalSlice;
