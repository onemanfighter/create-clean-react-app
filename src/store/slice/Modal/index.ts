export { default as createModalSlice } from './Modal.slice';
export type { ModalState, ModalStateSlice } from './types';
export { ModalOpenState, ModalIDs } from './types';
