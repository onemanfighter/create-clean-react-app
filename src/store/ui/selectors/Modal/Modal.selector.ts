import { UiStoreState } from '@uiStore';

const modalSelector = (state: UiStoreState) => ({
  openModal: state.Modal.openModal,
});

const modalDataSelector = (state: UiStoreState) => ({
  openState: state.Modal.modalOpenState,
  modalID: state.Modal.modalID,
  onModalClose: state.Modal.modalData?.onModalClose,
  resetModalState: state.Modal.resetModalState,
});

export { modalSelector, modalDataSelector };
