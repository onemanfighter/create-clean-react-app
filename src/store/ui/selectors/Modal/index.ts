export { modalSelector, modalDataSelector } from './Modal.selector';
