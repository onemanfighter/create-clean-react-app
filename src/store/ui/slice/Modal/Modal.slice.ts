import { UiStoreSlice } from '@uiStore';
import { ModalState, ModalStateSlice, ModalOpenState, ModalID } from './types';

const defaultModalState: ModalState = {
  modalID: ModalID.NONE,
  modalOpenState: ModalOpenState.CLOSE,
  modalData: undefined,
};

const createModalSlice: UiStoreSlice<ModalStateSlice> = (set) => ({
  ...defaultModalState,
  openModal: (id, onModalClose) => {
    set((state) => {
      state.Modal.modalData = { onModalClose };
      state.Modal.modalID = id;
      state.Modal.modalOpenState = ModalOpenState.OPEN;
    });
  },
  resetModalState: () => {
    setTimeout(() => {
      set((state) => {
        state.Modal.modalID = defaultModalState.modalID;
      });
    }, 200);

    set((state) => {
      state.Modal.modalOpenState = ModalOpenState.CLOSE;
      state.Modal.modalData = defaultModalState.modalData;
    });
  },
});

export default createModalSlice;
