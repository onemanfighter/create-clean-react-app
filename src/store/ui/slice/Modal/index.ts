export { default as createModalSlice } from './Modal.slice';
export type { ModalState, ModalStateSlice } from './types';
export { ModalOpenState, ModalID } from './types';
