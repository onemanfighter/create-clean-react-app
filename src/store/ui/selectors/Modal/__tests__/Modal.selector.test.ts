import { renderHook } from '@testing-library/react-hooks';
import { modalDataSelector, modalSelector } from '../Modal.selector';
import { ModalID, ModalOpenState, ModalState, appStore } from '@uiStore';

describe('Modal selector', () => {
  const onModalCloseMock = jest.fn();

  it('should return modal selector state and actions', () => {
    const modal = renderHook(() => appStore(modalSelector)).result.current;

    expect(modal).toMatchSnapshot();
  });

  it('should return modal state with initial value', () => {
    const { result } = renderHook(() => appStore(modalSelector));

    expect(result.current).toMatchSnapshot();
  });

  it('should return modal state on modal data set using openModal', () => {
    const { result: modalResult } = renderHook(() => appStore(modalSelector));
    const { result: dataResult } = renderHook(() =>
      appStore(modalDataSelector),
    );

    modalResult.current.openModal(ModalID.SEARCH, onModalCloseMock);

    expect(dataResult.current).toMatchSnapshot();
  });

  it('should reset and close modal state on modal data set using closeModal', () => {
    const modalData: ModalState = {
      modalID: ModalID.SEARCH,
      modalData: {
        onModalClose: onModalCloseMock,
      },
      modalOpenState: ModalOpenState.OPEN,
    };
    const { result: modalResult } = renderHook(() => appStore(modalSelector));
    const { result: dataResult } = renderHook(() =>
      appStore(modalDataSelector),
    );

    modalResult.current.openModal(ModalID.SEARCH, onModalCloseMock);

    expect(dataResult.current.modalID).toEqual(modalData.modalID);
    expect(dataResult.current.onModalClose).toEqual(
      modalData.modalData?.onModalClose,
    );

    dataResult.current.resetModalState();
    jest.runAllTimers();

    expect(dataResult.current.openState).toEqual(ModalOpenState.CLOSE);
    expect(dataResult.current.modalID).toEqual(ModalID.NONE);
    expect(dataResult.current.onModalClose).toEqual(undefined);
  });
});
