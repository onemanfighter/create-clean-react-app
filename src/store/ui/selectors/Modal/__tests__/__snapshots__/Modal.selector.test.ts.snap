// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal selector should return modal selector state and actions 1`] = `
{
  "openModal": [Function],
}
`;

exports[`Modal selector should return modal state on modal data set using openModal 1`] = `
{
  "modalID": "search",
  "onModalClose": [MockFunction],
  "openState": "OPEN",
  "resetModalState": [Function],
}
`;

exports[`Modal selector should return modal state with initial value 1`] = `
{
  "openModal": [Function],
}
`;
